### PR TITLE
disable method overwrite warnings if both functions are in Main

### LINKED ISSUE #18725: Julia 0.5 now complains about redefining methods
I recently upgraded to Julia 0.5 and now Julia raises a warning every time I redefine a function, which in my workflow happens **all the time**. I spent most of my day editing code and experimenting. My usual workflow looks like this:

``` julia
julia> include("foo.jl");
... look at the results, make edits ...

julia> include("foo.jl");
... look at the results, get an idea, make edits ...

julia> include("foo.jl");
... look at the results, get an idea, make edits ...

julia> include("foo.jl");
... look at the results, get an idea, make edits ...

julia> include("foo.jl");
... look at the results, get an idea, make edits ...

julia> include("foo.jl");
... look at the results, get an idea, make edits ...

```

Needless to say, every single time I `include("foo.jl")` I define all the methods in `foo.jl`. Now, with the new Julia 0.5 my output looks like this:

``` julia
julia>  include("main.jl");
WARNING: Method definition plot_obs() in module Main at /home/daniel/Science/C4/final-plots.jl:6
 overwritten at /home/daniel/Science/C4/final-plots.jl:6.
WARNING: Method definition plot_obs_mass_ratios() in module Main at /home/daniel/Science/
C4/final-plots.jl:158 overwritten at /home/daniel/Science/C4/final-plots.jl:158.
WARNING: Method definition plot_chosen_delta_as() in module Main at /home/daniel/Science/
C4/final-plots.jl:201 overwritten at /home/daniel/Science/C4/final-plots.jl:201.
WARNING: Method definition plot_chosen_delta_as(Any) in module Main at 
/home/daniel/Science/C4/final-plots.jl:201 overwritten at /home/daniel/Science/C4/final-plots.jl:201.
WARNING: Method definition parse_group(Any) in module Main at 
/home/daniel/Science/C4/simulations.jl:16 overwritten at /home/daniel/Science/C4/simulations.jl:16.
WARNING: Method definition plot_box(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/simulations.jl:147 overwritten at /home/daniel/Science/C4/simulations.jl:147.
WARNING: Method definition plot_theilsen() in module Main at 
/home/daniel/Science/C4/simulations.jl:194 overwritten at /home/daniel/Science/C4/simulations.jl:194.
WARNING: Method definition plot_three_planets() in module Main at 
/home/daniel/Science/C4/simulations.jl:280 overwritten at /home/daniel/Science/C4/simulations.jl:280.
WARNING: Method definition read_all() in module Main at /home/daniel/Science/C4/catalog.jl:5
 overwritten at /home/daniel/Science/C4/catalog.jl:5.
WARNING: Method definition read_mult() in module Main at /home/daniel/Science/C4/catalog.jl:60 
overwritten at /home/daniel/Science/C4/catalog.jl:60.
WARNING: Method definition plot_pairs(Any) in module Main at /home/daniel/Science/C4/catalog.jl:91
 overwritten at /home/daniel/Science/C4/catalog.jl:91.
WARNING: Method definition get_limits(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:20 overwritten at /home/daniel/Science/C4/gladman.jl:20.
WARNING: Method definition get_limits(Any, Any, Any, Any, Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:56 overwritten at /home/daniel/Science/C4/gladman.jl:56.
WARNING: Method definition draw_frame_gladman(Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:99 overwritten at /home/daniel/Science/C4/gladman.jl:99.
WARNING: Method definition draw_frame_chambers(Any, Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:123 overwritten at /home/daniel/Science/C4/gladman.jl:123.
WARNING: Method definition draw_frame_chambers(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:155 overwritten at /home/daniel/Science/C4/gladman.jl:155.
WARNING: Method definition draw_delta(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/gladman.jl:187 overwritten at /home/daniel/Science/C4/gladman.jl:187.
WARNING: Method definition final_orbits() in module Main at /home/daniel/Science/C4/gladman.jl:196
 overwritten at /home/daniel/Science/C4/gladman.jl:196.
WARNING: Method definition old_final_orbits() in module Main at 
/home/daniel/Science/C4/gladman.jl:221 overwritten at /home/daniel/Science/C4/gladman.jl:221.
WARNING: Method definition f(Any) in module Main at /home/daniel/Science/C4/hill.jl:24 overwritten
 at /home/daniel/Science/C4/hill.jl:49.
WARNING: Method definition df(Any) in module Main at /home/daniel/Science/C4/hill.jl:25 overwritten
 at /home/daniel/Science/C4/hill.jl:50.
WARNING: Method definition hill_limit(Any, Any, Any, Any, Any, Any) in module Main at
 /home/daniel/Science/C4/hill.jl:7 overwritten at /home/daniel/Science/C4/hill.jl:7.
WARNING: Method definition gladman_limit(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/hill.jl:81 overwritten at /home/daniel/Science/C4/hill.jl:81.
WARNING: Method definition chambers_limit(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/hill.jl:83 overwritten at /home/daniel/Science/C4/hill.jl:83.
WARNING: Method definition Rh_plot(Any, Any) in module Main at /home/daniel/Science/C4/hill.jl:88
 overwritten at /home/daniel/Science/C4/hill.jl:88.
WARNING: Method definition alpha_plot(Any, Any) in module Main at /home/daniel/Science/C4/hill.jl:99
 overwritten at /home/daniel/Science/C4/hill.jl:99.
WARNING: Method definition test_hill() in module Main at /home/daniel/Science/C4/hill.jl:108
 overwritten at /home/daniel/Science/C4/hill.jl:108.
WARNING: Method definition test_hill3() in module Main at /home/daniel/Science/C4/hill.jl:168
 overwritten at /home/daniel/Science/C4/hill.jl:168.
WARNING: Method definition test_hill2() in module Main at /home/daniel/Science/C4/hill.jl:245
 overwritten at /home/daniel/Science/C4/hill.jl:245.
WARNING: Method definition saveplot(Any) in module Main at /home/daniel/Science/C4/misc.jl:6
 overwritten at /home/daniel/Science/C4/misc.jl:6.
WARNING: Method definition theilsen(AbstractArray, AbstractArray) in module Main at
 /home/daniel/Science/C4/misc.jl:14 overwritten at /home/daniel/Science/C4/misc.jl:14.
WARNING: replacing docs for 'theilsen :: Tuple{AbstractArray,AbstractArray}' in module 'Main'.
WARNING: Method definition summarize_runs() in module Main at /home/daniel/Science/C4/misc.jl:38
 overwritten at /home/daniel/Science/C4/misc.jl:38.
WARNING: replacing docs for 'summarize_runs :: Tuple{}' in module 'Main'.
WARNING: Method definition plot_2D_kde(Any, Any, Any, Any) in module Main at
 /home/daniel/Science/C4/misc.jl:89 overwritten at /home/daniel/Science/C4/misc.jl:89.
WARNING: replacing docs for 'plot_2D_kde :: Tuple{Any,Any,Any,Any}' in module 'Main'.
WARNING: Method definition generate_synthetic_pop(Any, Any) in module Main at 
/home/daniel/Science/C4/misc.jl:112 overwritten at /home/daniel/Science/C4/misc.jl:112.
WARNING: Method definition generate_synthetic_pop(Any, Any, Any) in module Main at 
/home/daniel/Science/C4/misc.jl:112 overwritten at /home/daniel/Science/C4/misc.jl:112.
WARNING: replacing docs for 'generate_synthetic_pop :: Union{Tuple{Any,Any,Any},Tuple{Any,Any}}'
 in module 'Main'.
WARNING: Method definition generate_everything(DataFrames.DataFrame, Base.Dict) in module 
Main at /home/daniel/Science/C4/misc.jl:213 overwritten at /home/daniel/Science/C4/misc.jl:213.
WARNING: Method definition generate_everything(DataFrames.DataFrame, Base.Dict, Bool) in 
module Main at /home/daniel/Science/C4/misc.jl:213 overwritten at 
/home/daniel/Science/C4/misc.jl:213.
WARNING: Method definition plot_2D_obs(Any, Any, Any, Any) in module Main at 
/home/daniel/Science/C4/misc.jl:254 overwritten at /home/daniel/Science/C4/misc.jl:254.
WARNING: Method definition plot_2D_imf(Any, Any, Any, Any) in module Main at 
/home/daniel/Science/C4/misc.jl:264 overwritten at /home/daniel/Science/C4/misc.jl:264.
WARNING: Method definition plot_1D_kde(DataFrames.DataFrame, Base.Dict) in module Main 
at /home/daniel/Science/C4/misc.jl:274 overwritten at /home/daniel/Science/C4/misc.jl:274.
WARNING: Method definition plot_1D_kde(DataFrames.DataFrame, Base.Dict, Bool) in module
 Main at /home/daniel/Science/C4/misc.jl:274 overwritten at /home/daniel/Science/C4/misc.jl:274.
WARNING: B.q01D+32/23 -- expected 4 completions; got 5
WARNING: B.q01D+32/26 -- expected 4 completions; got 5
WARNING: B.q01D+32/27 -- expected 4 completions; got 5
WARNING: B.q01D+32/29 -- expected 4 completions; got 5
WARNING: B.q02D+34/14 is green but needs G*.end
```

Now all the legitimate warnings from my code are buried in a sea of useless warnings telling me that I overwrote a function again. Not only is this irritating, but I also think that it makes no sense in a functional language like Julia, where functions are just variables. Julia doesn't warn me every time that I change the value of a variable, right? Because that would be silly and annoying. But Julia functions are variables and if I define them in the form of a variable assignment, like `myadd = x,y -> x+y` Julia doesn't complain. So Julia is also being inconsistent in that some function redefinitions raise errors but others don't.

There is also no precedent that I know of for this kind of behaviour. In Python, when you modify a file and "import" the file again, you don't get a sea of warnings for every single function defined in that file. Ditto for Perl and the other high level languages that I can think of.

I would like to request that the old behaviour be restored. The new behaviour doesn't add anything of value and it creates a lot of hassle and frustration. Failing at that, I would like to at least have a setting that I can put my ~/.juliarc.jl file to turn off all these warnings.

### PR DESCRIPTION
**Pull request by KristofferC**
Implements @vtjnash suggestion in https://github.com/JuliaLang/julia/issues/18725#issuecomment-250502825

``` jl
julia> f(x) = 3;

julia> f(x) = 3;

julia> module F
       f(x) = 3
       f(x) = 3
       end;
WARNING: Method definition f(Any) in module F at REPL[2]:2 overwritten at REPL[2]:3.
```

fixes #18725

Stuff like https://github.com/tbreloff/Plots.jl/issues/508 or

``` jl

julia> f(g) = g();

julia> g() = 1;

julia> f(g)
1

julia> g() = 2

julia> f(g)
1
```

are of course a little bit scary to happen without a warning.

### CODE REVIEW

**Comment by tkelman** [comment]
can we make this only the case when `isinteractive` ?

File: test/compile.jl [hunk 1]
```diff
@@ -366,4 +366,30 @@ let module_name = string("a",randstring())
...
+    finally
+        close(STDERR)
+        redirect_stderr(olderr)
+    end
+    wait(t)
+
+    t = redirected_stderr("WARNING: Method definition f_overwrite_module(Any) in module F18725")
+    mktemp() do path, stream
+        try
+        include_string("""module F18725
```
Comment (tkelman) [inline test/compile.jl]: indent is off

File: test/compile.jl [hunk 2]
```diff
@@ -366,4 +366,30 @@ let module_name = string("a",randstring())
...
+        f_overwrite_main(x) = x
+        f_overwrite_main(x) = x
+    finally
+        close(STDERR)
+        redirect_stderr(olderr)
+    end
+    wait(t)
+
+    t = redirected_stderr("WARNING: Method definition f_overwrite_module(Any) in module F18725")
+    mktemp() do path, stream
```
Comment (tkelman) [inline test/compile.jl]: neither path or stream are used here, why the mktemp?

Comment (KristofferC) [inline test/compile.jl]: Accidentó

**Comment by KristofferC** [comment]
How should I test it  @tkelman. Start a new worker with the needed command flag?

**Comment by tkelman** [comment]
If that works? We don't do all that much testing of warnings, though that doesn't mean we shouldn't try to.

**Comment by KristofferC** [comment]
Regarding `isinteractive` there is a bit of a mismatch on the julia side vs c side:

```jl
julia> isinteractive()
true

julia> Base.JLOptions().isinteractive
0
```

because the function that starts the REPL sets the `isinteractive` variable here: https://github.com/JuliaLang/julia/blob/0287a9971ea4d740d2bc5e205423e2ae5bdf4124/base/client.jl#L346. I could of course just `jl_call` into julias `isinteractive()`. Would that be reasonable?

**Comment by vtjnash** [comment]
If we are going to make this conditional on `Base.isinteractive()`, it should probably be a push model – something like:
```julia
ccall(:jl_suppress_interactive_warnings, Cint, (Cint,), true)
```

```c
static int suppress_interactive_warnings;
JL_DLLEXPORT int jl_suppress_interactive_warnings(int nowarn)
{
    int last = suppress_interactive_warnings;
    suppress_interactive_warnings = nowarn;
    return last;
}
```

**Comment by JeffBezanson** [comment]
I think it would be better to simply add an option to disable the warning, at least until #265 is fixed. It's very hard to make the case that some method overwrites don't need warnings, and others do.

**Comment by KristofferC** [comment]
I agree.

**Comment by vtjnash** [comment]
> I think it would be better to simply add an option to disable the warning, at least until #265 is fixed. 

Either this should be a warning or it shouldn't. This isn't really doing much to warn for #265 occurrences, since it only handles one very specific case, so that seems like a red herring.

> It's very hard to make the case that some method overwrites don't need warnings, and others do.

Adding a flag is annoying since it's one more thing that will differ between master and worker processes. Jeff was the one who originally convinced me it's better to avoid options when we can just pick a sane default (such as warning only for code inside a module), or we could move this entirely to a linter.

**Comment by JeffBezanson** [comment]
We already have an option to disable deprecation warnings, so options of this kind are clearly considered acceptable at this point.

True, it does not catch all cases of #265, but the case in https://github.com/JuliaLang/julia/issues/18725#issuecomment-250908895 is especially bad, and I feel it's very helpful to give some indication of what might be wrong. Unlike other cases of #265, this one is actionable, since you can usually avoid overwriting methods (e.g. use different names instead of repeatedly redefining `f(x) = ` in a test suite).

This warning is perfectly fine except in the marginal case where you reload lots of definitions and find the excess text annoying. An option seems like the right way to handle that, so that you can temporarily silence the warnings to make it easier to see other messages.

**Comment by KristofferC** [comment]
How about showing it only once per session for Main Main overwrites?

**Pull request closed without merging**